src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2StreamTests.cs

```diff
@@ -3307,5 +3333,698 @@ public async Task ResponseBodyPipeCompleteWithExceptionThrows()
+
+            Assert.Single(_decodedHeaders);
+            Assert.Equal("Custom Value", _decodedHeaders["CustomName"]);
+        }
+
+        [Fact]
+        public async Task CompleteAsync_BeforeBodyStarted_WithTrailers_TruncatedContentLength_ThrowsAnd500()
+        {
+            var startingTcs = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
+            var appTcs = new TaskCompletionSource<int>(TaskCreationOptions.RunContinuationsAsynchronously);
+            var headers = new[]
+            {
+                new KeyValuePair<string, string>(HeaderNames.Method, "GET"),
+                new KeyValuePair<string, string>(HeaderNames.Path, "/"),
+                new KeyValuePair<string, string>(HeaderNames.Scheme, "http"),
+            };
+            await InitializeConnectionAsync(async context =>
+            {
+                try
+                {
+                    context.Response.OnStarting(() => { startingTcs.SetResult(0); return Task.CompletedTask; });
+                    var completionFeature = context.Features.Get<IHttpResponseCompletionFeature>();
+                    Assert.NotNull(completionFeature);
+
+                    context.Response.ContentLength = 25;
+                    context.Response.AppendTrailer("CustomName", "Custom Value");
+
+                    var ex = await Assert.ThrowsAsync<InvalidOperationException>(() => completionFeature.CompleteAsync().DefaultTimeout());
+                    Assert.Equal(CoreStrings.FormatTooFewBytesWritten(0, 25), ex.Message);
+
+                    Assert.True(startingTcs.Task.IsCompletedSuccessfully);
+                    Assert.False(context.Response.Headers.IsReadOnly);
+                    Assert.False(context.Features.Get<IHttpResponseTrailersFeature>().Trailers.IsReadOnly);
+
+                    appTcs.SetResult(0);
+                }
+                catch (Exception ex)
+                {
+                    appTcs.SetException(ex);
+                }
+            });
+
+            await StartStreamAsync(1, headers, endStream: true);
+
+            var headersFrame = await ExpectAsync(Http2FrameType.HEADERS,
+                withLength: 55,
+                withFlags: (byte)(Http2HeadersFrameFlags.END_HEADERS | Http2HeadersFrameFlags.END_STREAM),
+                withStreamId: 1);
+
+            await appTcs.Task;
+
+            await StopConnectionAsync(expectedLastStreamId: 1, ignoreNonGoAwayFrames: false);
+
+            _hpackDecoder.Decode(headersFrame.PayloadSequence, endHeaders: false, handler: this);
+
+            Assert.Equal(3, _decodedHeaders.Count);
+            Assert.Contains("date", _decodedHeaders.Keys, StringComparer.OrdinalIgnoreCase);
+            Assert.Equal("500", _decodedHeaders[HeaderNames.Status]);
+            Assert.Equal("0", _decodedHeaders[HeaderNames.ContentLength]);
+        }
+
+        [Fact]
+        public async Task CompleteAsync_AfterBodyStarted_SendsBodyWithEndStream()
```

Comment:
I believe you can write to the body pipe without starting the response. Should that be tested?

Reply:
Are you suggesting we test what happens when we call ResponseBodyPipe.GetMemory(), then ResponseBodyPipe.Advance(), then CompleteAsync(), and then finally ResponseBodyPipe.FlushAsync()?

I expect FlushAsync() would throw and InvalidOperationException in that case. That seems like a reasonable test.

Reply:
I was thinking:

1. GetMemory
2. Advance
3. CompleteAsync
4. Assert that written data is sent (CompleteAsync would match FlushAsync in this regard)

FlushAsync erroring after CompleteAsync is probably already tested.

Reply:
Agree with @JamesNK. That code is spicy and deserves a test.